i6_models/parts/factored_hybrid/diphone.py

```diff
@@ -0,0 +1,145 @@
+
+
+@dataclass
+class FactoredDiphoneBlockV1Config(ModelConfiguration):
+    """
+    Attributes:
+        context_mix_mlp_dim: inner dimension of the context mixing MLP layers
+        context_mix_mlp_num_layers: how many hidden layers on the MLPs there should be
+        dropout: dropout probabilty
+        left_context_embedding_dim: embedding dimension of the left context
+            values. Good choice is in the order of num_contexts.
+        num_contexts: the number of raw phonemes/acoustic contexts
+        num_hmm_states_per_phone: the number of HMM states per phoneme
+        num_inputs: input dimension of the output block, must match w/ output dimension
+            of main encoder (e.g. Conformer)
+        phoneme_state_class: the phoneme state augmentation to apply
+        activation: activation function to use in the context mixing MLP.
+    """
+
+    left_context_embedding_dim: int
+    num_contexts: int
+    num_hmm_states_per_phone: int
+    boundary_class: Union[int, BoundaryClassV1]
+
+    activation: Callable[[], nn.Module]
+    context_mix_mlp_dim: int
+    context_mix_mlp_num_layers: int
+    dropout: float
+    num_inputs: int
+
+    def __post_init__(self) -> None:
+        super().__post_init__()
+
+        assert self.context_mix_mlp_dim > 0
+        assert self.context_mix_mlp_num_layers > 0
+        assert self.num_contexts > 0
+        assert self.num_hmm_states_per_phone > 0
+        assert self.num_inputs > 0
+        assert self.left_context_embedding_dim > 0
+        assert 0.0 <= self.dropout <= 1.0, "dropout must be a probability"
+
+
+class FactoredDiphoneBlockV1(nn.Module):
+    """
+    Diphone FH model output block.
+
+    Consumes the output h(x) of a main encoder model and computes factored or joint
+    output logits/probabilities for p(c|l,x) and p(l|x).
+    """
+
+    def __init__(self, cfg: FactoredDiphoneBlockV1Config):
+        super().__init__()
+
+        self.boundary_class = cfg.boundary_class
+        self.num_contexts = cfg.num_contexts
+        self.num_hmm_states_per_phone = cfg.num_hmm_states_per_phone
+
+        self.num_center = get_center_dim(self.num_contexts, self.num_hmm_states_per_phone, self.boundary_class)
+        self.num_diphone = self.num_center * self.num_contexts
+
+        self.left_context_encoder = get_mlp(
+            num_input=cfg.num_inputs,
+            num_output=cfg.num_contexts,
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+        self.left_context_embedding = nn.Embedding(cfg.num_contexts, cfg.left_context_embedding_dim)
+        self.center_encoder = get_mlp(
+            num_input=cfg.num_inputs + cfg.left_context_embedding_dim,
+            num_output=self.num_center,
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
```

Comment:
Is this intentionally using the same dim/num_layers as the (left-)context MLP? Maybe rename those parameters to a name that reflects that both left-context and center have these dims?

Reply:
My intuition says the context mixing MLPs themselves don't have a big impact on model performance as long as they're sized sufficiently to perform their context mixing job, but maybe for the higher order contexts the size becomes more relevant due to the additional dependencies. Then again, this is also just intuition.

@Marvin84 Do you have any experience about the difference the sizing of these layers makes on the overall WER? The dim and number of layers here are the same because (I think) we never really experimented around much with them.

Reply:
It would certainly help if it turned out that these layers could even be sized down, and thereby speed up the score computation vs. a CART model (joint FH diphone is slower than triphone CART due to batching the context mixing MLPs over all contexts, whereas the CART just upsizes the model output via one linear layer to the final output dim).

Reply:
I experimented with BLSTM based models, and in that case one layer against two layers had max `0.2%` absolute WER degradation on Hub500 (trained on SWB300h). Moreover, lower embed dimension could lead to absolute `0.1%` WER degradation. 
However, I noticed that for from-scratch Conformed based alignment models (monophone PHMM) having or not having MLP layers matter a lot. I have not checked the effect of number of MLP layers for Viterbi training with fixed path, though.

> speed up the score computation vs. a CART model 

How much speedup do you expect to get?

Reply:
<img width="672" alt="image" src="https://github.com/user-attachments/assets/21970f83-7d47-4624-93b9-0e33dc6ba332">

Upper line is triphone CART, lower line diphone FH. Would be nice if the TF cost for both was at least equal. It's unlikely we can get FH any better, though.

Reply:
> Upper line is triphone CART, lower line diphone FH.

should we not compare explicit full-diphone to joint forwarding of factored diphone?

Reply:
In any case I think making this any more configurable can be postponed to a V2.

> should we not compare explicit full-diphone to joint forwarding of factored diphone?

Technically, yes. There won't be a noticeable difference to a triphone CART model wrt. the NN cost, however, as the size of the output layers is almost equal (at least for Librispeech, where it's 12k CART vs 11.5k diphone).